File: tape/sdk/typescript/src/outbox_reactor.ts
// Outbox-reactor daemon — TS counterpart of Python's `tape.reactors.outbox`.
//
// One pass:
//   list effects to dispatch (PENDING + OUTBOX + due)
//   for each:
//     claim (atomic CAS lease)
//     look up the connector
//     dispatch through it
//     record result:
//       confirmed → completeEffect(CONFIRMED)
//       failed    → recordDispatchAttempt(next_at = backoff) → eventually FAILED
//       unknown   → recordDispatchAttempt(next_at = 0) → status UNKNOWN
//                   (the reconciler resolves; do NOT blindly retry —
//                    that is the entire safety claim for non-idempotent
//                    upstreams)
//
// Safety: the server's CAS on claim_effect_dispatch enforces non-blind-retry;
// this reactor double-checks (refuses to act on anything not PENDING after
// the claim).

import * as os from 'node:os';
import {
  TapeClient, DEFAULT_URL,
  EffectStatus, EffectSemantics, EffectResolution,
} from './client.ts';
import {
  CONNECTORS, ConnectorRegistry, type Connector, type EffectRecord, type DispatchResult,
} from './connectors/index.ts';

export interface OutboxReactorOptions {
  /** TapeClient — pass an existing one or let the loop create its own. */
  client?: TapeClient;
  /** URL when constructing a new client. Honoured iff `client` is undefined. */
  url?: string;
  /** Restrict to one connector name (matches @outboxTool({ connector }) on the tool). */
  connector?: string;
  /** Custom registry; defaults to CONNECTORS (the process-global). */
  registry?: ConnectorRegistry;
  /** Identity recorded as `dispatch_claimed_by`. */
  claimer?: string;
  /** Page size. */
  limit?: number;
  /** Give up on a connector failure after N attempts (then mark FAILED). */
  dispatchMaxAttempts?: number;
}

export interface OutboxOutcome {
  runId: string;
  idempotencyKey: string;
  connector: string;
  tool: string;
  status: 'confirmed' | 'unknown' | 'failed' | 'retry-scheduled' | 'skipped' | 'error';
  reason?: string;
  externalRef?: string;
  error?: string;
  nextAtMs?: number;
  attempts?: number;
}

function claimerId(): string {
  return process.env.TAPE_DISPATCH_CLAIMER ?? `${os.hostname()}:${process.pid}`;
}

function backoffMs(attempt: number, baseS = 1.0, maxS = 60.0): number {
  const delayS = Math.min(baseS * Math.pow(2, Math.max(attempt - 1, 0)), maxS);
  return Math.floor(delayS * 1000);
}

function toConnectorEffect(eff: any): EffectRecord {
  let payload: unknown = eff.requestJson ?? '';
  if (typeof payload === 'string' && payload.length > 0) {
    try { payload = JSON.parse(payload); } catch { /* leave string */ }
  }
  return {
    runId: eff.runId,
    idempotencyKey: eff.idempotencyKey,
    toolName: eff.toolName,
    connector: eff.connector,
    payload,
    businessKey: eff.businessKey ?? '',
    attempt: (eff.dispatchAttempts ?? 0) + 1,
    semantics: eff.semantics === EffectSemantics.NON_IDEMPOTENT ? 'non_idempotent' : 'idempotent',
    tenantId: eff.tenantId ?? '',
    appName: eff.appName ?? '',
  };
}

export async function dispatchOne(
  eff: any,
  opts: Required<Pick<OutboxReactorOptions, 'client' | 'claimer' | 'dispatchMaxAttempts'>>
    & { registry: ConnectorRegistry },
): Promise<OutboxOutcome> {
  const out: OutboxOutcome = {
    runId: eff.runId,
    idempotencyKey: eff.idempotencyKey,
    connector: eff.connector,
    tool: eff.toolName,
    status: 'skipped',
  };

  if (!opts.registry.has(eff.connector)) {
    out.reason = `no connector registered: '${eff.connector}'`;
    return out;
  }
  const connector: Connector = opts.registry.get(eff.connector);

  const claim: any = await opts.client.claimEffectDispatch({
    runId: eff.runId, idempotencyKey: eff.idempotencyKey,
    claimer: opts.claimer, leaseTtlMs: 60_000,
  });
  if (!claim.acquired) { out.reason = 'lease contended'; return out; }

  const cur = claim.effect;
  if (cur.status !== EffectStatus.PENDING) {
    out.reason = `unexpected status after claim: ${cur.status}`;
    return out;
  }

  const isNonIdem = cur.semantics === EffectSemantics.NON_IDEMPOTENT;
  let result: DispatchResult;
  try {
    result = await connector.dispatch(toConnectorEffect(cur));
  } catch (ex: any) {
    if (isNonIdem) {
      await opts.client.recordDispatchAttempt({
        runId: cur.runId, idempotencyKey: cur.idempotencyKey,
        error: `connector raised: ${ex?.name ?? 'Error'}: ${ex?.message ?? String(ex)}`,
        nextDispatchAtMs: 0,
      });
      out.status = 'unknown'; out.error = String(ex);
      return out;
    }
    const attempts = (cur.dispatchAttempts ?? 0) + 1;
    const nextAt = Date.now() + backoffMs(attempts);
    await opts.client.recordDispatchAttempt({
      runId: cur.runId, idempotencyKey: cur.idempotencyKey,
      error: `connector raised: ${ex?.name ?? 'Error'}: ${ex?.message ?? String(ex)}`,
      nextDispatchAtMs: nextAt,
    });
    out.status = 'retry-scheduled'; out.error = String(ex); out.nextAtMs = nextAt; out.attempts = attempts;
    return out;
  }

  if (result.outcome === 'confirmed') {
    const responseBody = result.response ?? {};
    const responseJson = JSON.stringify({
      external_ref: result.dispatchId ?? '',
      ...(typeof responseBody === 'object' && responseBody !== null ? responseBody : { value: responseBody }),
    });
    await opts.client.completeEffect({
      runId: cur.runId, idempotencyKey: cur.idempotencyKey,
      status: EffectStatus.CONFIRMED, responseJson,
    });
    out.status = 'confirmed'; out.externalRef = result.dispatchId ?? '';
    return out;
  }

  if (result.outcome === 'unknown') {
    await opts.client.recordDispatchAttempt({
      runId: cur.runId, idempotencyKey: cur.idempotencyKey,
      error: JSON.stringify(result.error ?? { reason: 'ack lost' }),
      nextDispatchAtMs: 0,
    });
    out.status = 'unknown';
    return out;
  }

  // failed / pending → retry or terminal-FAILED
  const attempts = (cur.dispatchAttempts ?? 0) + 1;
  if (attempts >= opts.dispatchMaxAttempts) {
    await opts.client.recordExternalObservation({
      runId: cur.runId, idempotencyKey: cur.idempotencyKey,
      resolution: EffectResolution.FAILED,
      errorJson: JSON.stringify({ final: true, attempts, last: result.error ?? null }),
    });
    out.status = 'failed'; out.attempts = attempts;
    return out;
  }
  const nextAt = (result.retryAfterMs ?? 0) > 0
    ? Date.now() + result.retryAfterMs!
    : Date.now() + backoffMs(attempts);
  await opts.client.recordDispatchAttempt({
    runId: cur.runId, idempotencyKey: cur.idempotencyKey,
    error: JSON.stringify(result.error ?? {}),
    nextDispatchAtMs: nextAt,
  });
  out.status = 'retry-scheduled'; out.nextAtMs = nextAt; out.attempts = attempts;
  return out;
}

export async function outboxDispatchOnce(opts: OutboxReactorOptions = {}): Promise<OutboxOutcome[]> {
  const ownsClient = opts.client === undefined;
  const client = opts.client ?? new TapeClient(opts.url ?? DEFAULT_URL);
  const registry = opts.registry ?? CONNECTORS;
  const claimer = opts.claimer || claimerId();
  const limit = opts.limit ?? 200;
  const dispatchMaxAttempts = opts.dispatchMaxAttempts ?? 5;
  const outcomes: OutboxOutcome[] = [];
  try {
    const resp: any = await client.listEffectsToDispatch({ connector: opts.connector ?? '', limit });
    for (const eff of resp.effects ?? []) {
      try {
        outcomes.push(await dispatchOne(eff, { client, registry, claimer, dispatchMaxAttempts }));
      } catch (ex: any) {
        outcomes.push({
          runId: eff.runId, idempotencyKey: eff.idempotencyKey,
          connector: eff.connector, tool: eff.toolName,
          status: 'error', error: String(ex),
        });
      }
    }
  } finally {
    if (ownsClient) client.close();
  }
  return outcomes;
}

export interface RunOutboxOptions extends OutboxReactorOptions {
  intervalMs?: number;
  once?: boolean;
  onTick?: (outcomes: OutboxOutcome[]) => void;
}

export async function runOutboxDispatcher(opts: RunOutboxOptions = {}): Promise<void> {
  const ownsClient = opts.client === undefined;
  const client = opts.client ?? new TapeClient(opts.url ?? DEFAULT_URL);
  const intervalMs = opts.intervalMs ?? 1000;
  try {
    // eslint-disable-next-line no-constant-condition
    while (true) {
      let outcomes: OutboxOutcome[] = [];
      try {
        outcomes = await outboxDispatchOnce({ ...opts, client });
      } catch (ex: any) {
        process.stderr.write(`[tape outbox] tick error: ${ex?.message ?? ex}\n`);
      }
      opts.onTick?.(outcomes);
      if (opts.once) return;
      await new Promise((r) => setTimeout(r, intervalMs));
    }
  } finally {
    if (ownsClient) client.close();
  }
}

File: tape/sdk/java/src/main/java/dev/tape/reactors/OutboxReactor.java
package dev.tape.reactors;

import com.google.gson.Gson;
import dev.tape.TapeClient;
import dev.tape.connectors.Connector;
import dev.tape.connectors.ConnectorRegistry;
import dev.tape.proto.*;

import java.lang.management.ManagementFactory;
import java.net.InetAddress;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Outbox-reactor — Java counterpart of Python's {@code tape.reactors.outbox}.
 *
 * <p>One pass: list effects to dispatch (PENDING+OUTBOX+due), then for each:
 * claim (atomic CAS lease), look up the connector, dispatch through it, and
 * record the result:
 *
 * <ul>
 *   <li>confirmed → {@link TapeClient#completeEffect}({@code CONFIRMED})</li>
 *   <li>failed → {@code recordDispatchAttempt(next_at = backoff)} (eventually FAILED)</li>
 *   <li>unknown → {@code recordDispatchAttempt(next_at = 0)} (status UNKNOWN;
 *       the reconciler resolves — do not blindly retry; that is the entire
 *       safety claim for non-idempotent upstreams)</li>
 * </ul>
 *
 * <p>Safety: the server's CAS on {@code claimEffectDispatch} enforces
 * non-blind-retry; this reactor double-checks (refuses to act if status is not
 * {@code PENDING} after the claim).
 */
public final class OutboxReactor {

    private static final Gson GSON = new Gson();

    private OutboxReactor() {}

    /** Per-effect outcome of one dispatch attempt. */
    public static final class Outcome {
        public String runId;
        public String idempotencyKey;
        public String connector;
        public String tool;
        /** confirmed | unknown | failed | retry-scheduled | skipped | error */
        public String status;
        public String reason;
        public String externalRef;
        public String error;
        public long   nextAtMs;
        public int    attempts;

        public Map<String, Object> toMap() {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("run_id", runId);
            m.put("idempotency_key", idempotencyKey);
            m.put("connector", connector);
            m.put("tool", tool);
            m.put("status", status);
            if (reason != null)      m.put("reason", reason);
            if (externalRef != null) m.put("external_ref", externalRef);
            if (error != null)       m.put("error", error);
            if (nextAtMs > 0)        m.put("next_at_ms", nextAtMs);
            if (attempts > 0)        m.put("attempts", attempts);
            return m;
        }
    }

    /** Dispatcher options. */
    public static final class Options {
        public String connector = "";          // restrict to one connector name
        public ConnectorRegistry registry;     // defaults to ConnectorRegistry.DEFAULT
        public String claimer = "";            // identity for dispatch_claimed_by
        public long   limit = 200;
        public int    dispatchMaxAttempts = 5;

        public Options connector(String v) { this.connector = v == null ? "" : v; return this; }
        public Options registry(ConnectorRegistry v) { this.registry = v; return this; }
        public Options claimer(String v) { this.claimer = v == null ? "" : v; return this; }
        public Options limit(long v) { this.limit = v <= 0 ? 200 : v; return this; }
        public Options dispatchMaxAttempts(int v) { this.dispatchMaxAttempts = v <= 0 ? 5 : v; return this; }
    }

    /** Long-lived dispatcher options. */
    public static final class RunOptions {
        public Options outbox = new Options();
        public long    intervalMs = 1000;
        public boolean once = false;
        public java.util.function.Consumer<List<Outcome>> onTick;
    }

    private static String claimerId() {
        String env = System.getenv("TAPE_DISPATCH_CLAIMER");
        if (env != null && !env.isEmpty()) return env;
        String host = "unknown";
        try { host = InetAddress.getLocalHost().getHostName(); } catch (Exception ignore) {}
        String pid = ManagementFactory.getRuntimeMXBean().getName(); // "<pid>@host"
        int at = pid.indexOf('@'); if (at > 0) pid = pid.substring(0, at);
        return host + ":" + pid;
    }

    private static long backoffMs(int attempt) {
        double delayS = Math.min(1.0 * Math.pow(2.0, Math.max(attempt - 1, 0)), 60.0);
        return (long) (delayS * 1000.0);
    }

    @SuppressWarnings("unchecked")
    private static Connector.Effect toConnectorEffect(EffectRecord eff) {
        Connector.Effect e = new Connector.Effect();
        e.runId = eff.getRunId();
        e.idempotencyKey = eff.getIdempotencyKey();
        e.toolName = eff.getToolName();
        e.connector = eff.getConnector();
        e.businessKey = eff.getBusinessKey();
        e.attempt = eff.getDispatchAttempts() + 1;
        e.semantics = (eff.getSemantics() == EffectSemantics.EFFECT_SEMANTICS_NON_IDEMPOTENT)
                ? "non_idempotent" : "idempotent";
        String body = eff.getRequestJson();
        if (body != null && !body.isEmpty()) {
            try { e.payload = GSON.fromJson(body, Map.class); }
            catch (Exception ignore) {
                Map<String, Object> m = new HashMap<>(); m.put("raw", body); e.payload = m;
            }
        } else {
            e.payload = new HashMap<>();
        }
        return e;
    }

    /** Run a single effect through its connector. Returns the per-effect outcome. */
    public static Outcome dispatchOne(TapeClient client, EffectRecord eff, Options opt) {
        Outcome out = new Outcome();
        out.runId = eff.getRunId();
        out.idempotencyKey = eff.getIdempotencyKey();
        out.connector = eff.getConnector();
        out.tool = eff.getToolName();
        out.status = "skipped";

        ConnectorRegistry reg = (opt.registry != null) ? opt.registry : ConnectorRegistry.DEFAULT;
        if (!reg.has(eff.getConnector())) {
            out.reason = "no connector registered: '" + eff.getConnector() + "'";
            return out;
        }
        Connector connector = reg.get(eff.getConnector());

        String claimer = (opt.claimer == null || opt.claimer.isEmpty()) ? claimerId() : opt.claimer;
        ClaimEffectDispatchResponse claim = client.claimEffectDispatch(
                eff.getRunId(), eff.getIdempotencyKey(), claimer, 60_000L);
        if (!claim.getAcquired()) { out.reason = "lease contended"; return out; }

        EffectRecord cur = claim.getEffect();
        if (cur.getStatus() != EffectStatus.EFFECT_STATUS_PENDING) {
            out.reason = "unexpected status after claim: " + cur.getStatus();
            return out;
        }

        boolean isNonIdem = (cur.getSemantics() == EffectSemantics.EFFECT_SEMANTICS_NON_IDEMPOTENT);

        Connector.Result result;
        try {
            result = connector.dispatch(toConnectorEffect(cur));
        } catch (Exception ex) {
            String msg = "connector raised: " + ex.getClass().getSimpleName() + ": " + ex.getMessage();
            if (isNonIdem) {
                client.recordDispatchAttempt(cur.getRunId(), cur.getIdempotencyKey(), msg, 0L);
                out.status = "unknown"; out.error = ex.toString();
                return out;
            }
            int attempts = cur.getDispatchAttempts() + 1;
            long nextAt = System.currentTimeMillis() + backoffMs(attempts);
            client.recordDispatchAttempt(cur.getRunId(), cur.getIdempotencyKey(), msg, nextAt);
            out.status = "retry-scheduled"; out.error = ex.toString();
            out.nextAtMs = nextAt; out.attempts = attempts;
            return out;
        }

        switch (result.outcome) {
            case CONFIRMED: {
                Map<String, Object> body = new LinkedHashMap<>();
                body.put("external_ref", result.dispatchId == null ? "" : result.dispatchId);
                if (result.response instanceof Map) {
                    body.putAll((Map<String, Object>) result.response);
                } else if (result.response != null) {
                    body.put("value", result.response);
                }
                client.completeEffect(cur.getRunId(), cur.getIdempotencyKey(),
                        EffectStatus.EFFECT_STATUS_CONFIRMED, GSON.toJson(body), "");
                out.status = "confirmed";
                out.externalRef = result.dispatchId == null ? "" : result.dispatchId;
                return out;
            }
            case UNKNOWN: {
                Map<String, Object> err = new LinkedHashMap<>();
                err.put("reason", "ack lost");
                if (result.error != null && !result.error.isEmpty()) err.put("error", result.error);
                client.recordDispatchAttempt(cur.getRunId(), cur.getIdempotencyKey(),
                        GSON.toJson(err), 0L);
                out.status = "unknown";
                return out;
            }
            default: {
                int attempts = cur.getDispatchAttempts() + 1;
                if (attempts >= opt.dispatchMaxAttempts) {
                    Map<String, Object> err = new LinkedHashMap<>();
                    err.put("final", true);
                    err.put("attempts", attempts);
                    err.put("last", result.error);
                    client.recordExternalObservation(cur.getRunId(), cur.getIdempotencyKey(),
                            EffectResolution.EFFECT_RESOLUTION_FAILED, "", "", GSON.toJson(err), "");
                    out.status = "failed"; out.attempts = attempts;
                    return out;
                }
                long nextAt = (result.retryAfterMs > 0)
                        ? System.currentTimeMillis() + result.retryAfterMs
                        : System.currentTimeMillis() + backoffMs(attempts);
                Map<String, Object> err = new LinkedHashMap<>();
                if (result.error != null && !result.error.isEmpty()) err.put("error", result.error);
                client.recordDispatchAttempt(cur.getRunId(), cur.getIdempotencyKey(),
                        GSON.toJson(err), nextAt);
                out.status = "retry-scheduled";
                out.nextAtMs = nextAt; out.attempts = attempts;
                return out;
            }
        }
    }

    /** One pass of the outbox dispatcher. */
    public static List<Outcome> dispatchOnce(TapeClient client, Options opt) {
        ListEffectsToDispatchResponse resp = client.listEffectsToDispatch(opt.connector, opt.limit, 0L);
        List<Outcome> outs = new ArrayList<>(resp.getEffectsCount());
        for (EffectRecord e : resp.getEffectsList()) {
            try { outs.add(dispatchOne(client, e, opt)); }
            catch (Exception ex) {
                Outcome o = new Outcome();
                o.runId = e.getRunId(); o.idempotencyKey = e.getIdempotencyKey();
                o.connector = e.getConnector(); o.tool = e.getToolName();
                o.status = "error"; o.error = ex.toString();
                outs.add(o);
            }
        }
        return outs;
    }

    /** Long-lived loop. Returns when {@code once} is true or the AtomicBoolean is cleared. */
    public static void run(TapeClient client, RunOptions opt, AtomicBoolean keepGoing) throws InterruptedException {
        if (opt.intervalMs <= 0) opt.intervalMs = 1000;
        while (true) {
            List<Outcome> outs;
            try { outs = dispatchOnce(client, opt.outbox); }
            catch (Exception ex) {
                System.err.println("[tape outbox] tick error: " + ex);
                outs = new ArrayList<>();
            }
            if (opt.onTick != null) opt.onTick.accept(outs);
            if (opt.once || (keepGoing != null && !keepGoing.get())) return;
            Thread.sleep(opt.intervalMs);
        }
    }
}
